catalog/core/catalog-core-impl/filter-proxy/src/main/java/ddf/catalog/filter/proxy/adapter/GeotoolsFilterAdapterImpl.java

```diff
@@ -130,42 +130,52 @@ public Object visit(Divide expression, Object delegate) {
         throw new UnsupportedOperationException(Divide.NAME + " expression not supported.");
     }
 
+    @Override
     public Object visit(Function expression, Object delegate) {
         if (expression == null) {
             throw new UnsupportedOperationException(
                     Function.class.getSimpleName() + " must not be null.");
         }
+        List<Expression> parameters = expression.getParameters();
 
-        if (expression.getName()
-                .equals(ProximityFunction.NAME.getName())) {
-            List<Expression> parameters = new ArrayList<>(expression.getParameters());
-            if (parameters.size() == ProximityFunction.NUM_PARAMETERS) {
-
-                String propertyName = parameters.remove(0)
-                        .toString();
-                int distance;
-                try {
-                    distance = Integer.parseInt(parameters.remove(0)
-                            .toString());
-                } catch (NumberFormatException e) {
-                    throw new IllegalArgumentException("Distance parameter was not valid");
-                }
-                String searchTerm = parameters.remove(0)
-                        .toString();
-                return new ProximityArguments(propertyName, distance, searchTerm);
-            } else {
-                throw new IllegalArgumentException("Not enough arguments for ProximityFunction");
-            }
+        int argCount = expression.getFunctionName()
+                .getArgumentCount();
+
+        if (parameters == null || parameters.size() != argCount) {
+            throw new UnsupportedOperationException(
+                    Function.class.getSimpleName() + " requires " + argCount + " arguments.");
+        }
+        //unwrap the function arguments and return them
+        List<Object> ret = new ArrayList<>();
+        for (int i = 0; i < argCount; i++) {
+            Class<?> type = expression.getFunctionName()
+                    .getArguments()
+                    .get(i)
+                    .getType();
+            Expression arg = parameters.get(i);
+            ret.add(arg.accept(this, type)); //the type would only be used by literals
         }
 
-        return expression;
+        return ret;
     }
 
-    public Object visit(Literal expression, Object delegate) {
+    @Override
+
+    public Object visit(Literal expression, Object clazz) {
         if (expression.getValue() == null) {
             throw new UnsupportedOperationException(
                     Literal.class.getSimpleName() + " value must not be null.");
         }
+        //try and get the object as the requested class otherwise just return the object
+        if (clazz instanceof Class) {
```

Comment:
Its to document what this visitor expects as an extraData object for other developers since the ExpressionVisitor interface requires that it be an object.